web/pdf_find_controller.js

```diff
@@ -358,136 +449,80 @@ class PDFFindController {
-      matchesLength.push(originalQueryLen);
+    const diffs = this._pageDiffs[pageIndex];
+    let match;
+    while ((match = query.exec(pageContent)) !== null) {
+      const [matchPos, matchLen] = getOriginalIndex(
+        diffs,
+        match.index,
+        match[0].length
+      );
+      matches.push(matchPos);
+      matchesLength.push(matchLen);
     }
     this._pageMatches[pageIndex] = matches;
     this._pageMatchesLength[pageIndex] = matchesLength;
   }
 
-  _calculateWordMatch(query, pageIndex, pageContent, pageDiffs, entireWord) {
-    const matchesWithLength = [];
+  _convertToRegExpString(query) {
+    const { entireWord, matchDiacritics } = this._state;
 
-    // Divide the query into pieces and search for text in each piece.
-    const queryArray = query.match(/\S+/g);
-    for (let i = 0, len = queryArray.length; i < len; i++) {
-      const subquery = queryArray[i];
-      const subqueryLen = subquery.length;
+    // Escape characters like *+?... to not interfer with regexp syntax.
+    query = query.replace(escapeRegExp, "\\$&");
 
-      let matchIdx = -subqueryLen;
-      while (true) {
-        matchIdx = pageContent.indexOf(subquery, matchIdx + subqueryLen);
-        if (matchIdx === -1) {
-          break;
-        }
-        if (
-          entireWord &&
-          !this._isEntireWord(pageContent, matchIdx, subqueryLen)
-        ) {
-          continue;
+    if (matchDiacritics) {
+      // aX musn't match aXY.
+      query = `${query}(?=[^\\p{Mn}])`;
+    } else {
+      query = query.replace(prepareNoDiacriticsRegExp, (match, p1) => {
+        if (p1) {
+          // Diacritics are removed.
+          return "";
         }
-        const originalMatchIdx = getOriginalIndex(matchIdx, pageDiffs),
-          matchEnd = matchIdx + subqueryLen - 1,
-          originalQueryLen =
-            getOriginalIndex(matchEnd, pageDiffs) - originalMatchIdx + 1;
-
-        // Other searches do not, so we store the length.
-        matchesWithLength.push({
-          match: originalMatchIdx,
-          matchLength: originalQueryLen,
-          skipped: false,
-        });
-      }
+        // A letter has been matched and it can be followed by any diacritics
+        // in normalized text.
+        return `${match}\\p{Mn}*`;
+      });
     }
 
-    // Prepare arrays for storing the matches.
-    this._pageMatchesLength[pageIndex] = [];
-    this._pageMatches[pageIndex] = [];
+    // Replace spaces by \s+ to be sure to match any spaces.
+    // We must do it after the if (matchDiacritcs) block to avoid
+    // wrong things with the "s".
+    query = query.replace(whitespacesRegExp, "\\s+");
 
-    // Sort `matchesWithLength`, remove intersecting terms and put the result
-    // into the two arrays.
-    this._prepareMatches(
-      matchesWithLength,
-      this._pageMatches[pageIndex],
-      this._pageMatchesLength[pageIndex]
-    );
+    if (entireWord) {
+      query = `\\b${query}\\b`;
+    }
```

Comment:
Are you *absolutely sure* that this catches *all* of the cases that the old `getCharacterType` functionality did?
Instinctively this does "feel" a lot less complete than the previous code, which compared different character classes in some detail; that implementation was borrowed from https://searchfox.org/mozilla-central/rev/083983b7f09b00cdfe4f70396e39ea4f8a1735e1/intl/lwbrk/WordBreaker.cpp#44-120

Reply:
I thought `\b` was for unicode too... so I'll fix that, thanks for catching that.